packages/vite-plugin-cloudflare/src/websockets.ts

```diff
@@ -0,0 +1,88 @@
+import ws from 'ws';
+import { UNKNOWN_HOST } from './shared';
+import type { Fetcher } from '@cloudflare/workers-types/experimental';
+import type { ReplaceWorkersTypes } from 'miniflare';
+import type { IncomingMessage } from 'node:http';
+import type { Duplex } from 'node:stream';
+import type * as vite from 'vite';
+
+export function handleWebSocket(
+	httpServer: vite.HttpServer,
+	fetcher: ReplaceWorkersTypes<Fetcher>['fetch'],
+	logger: vite.Logger,
+) {
+	const nodeWebSocket = new ws.Server({ noServer: true });
+
+	httpServer.on(
+		'upgrade',
+		async (request: IncomingMessage, socket: Duplex, head: Buffer) => {
+			const url = new URL(request.url ?? '', UNKNOWN_HOST);
+
+			// Ignore Vite HMR WebSockets
+			if (request.headers['sec-websocket-protocol']?.startsWith('vite')) {
+				return;
+			}
+
+			const headers = new Headers();
+
+			for (const [key, value] of Object.entries(request.headers)) {
+				if (typeof value === 'string') {
+					headers.append(key, value);
+				} else if (Array.isArray(value)) {
+					for (const item of value) {
+						headers.append(key, item);
+					}
+				}
+			}
```

Comment:
this could probably be isolated in its own function no? (to make this function simpler to read/follow)

Reply:
[ec83e5b](https://github.com/flarelabs-net/vite-plugin-cloudflare/pull/128/commits/ec83e5bbc06d520275333bb946fd26b7ab8f064a)